posthog/api/organization.py

```diff
@@ -0,0 +1,96 @@
+from rest_framework import (
+    exceptions,
+    mixins,
+    permissions,
+    request,
+    response,
+    serializers,
+    status,
+    viewsets,
+)
+
+from posthog.models import Organization, OrganizationMembership, Team
+from posthog.permissions import CREATE_METHODS, OrganizationAdminWritePermissions, OrganizationMemberPermissions
+
+
+class PremiumMultiorganizationPermissions(permissions.BasePermission):
+    """Require user to have all necessary premium features on their plan for create access to the endpoint."""
+
+    message = (
+        "You must upgrade your PostHog plan to be able to create and administrate multiple projects in an organization."
+    )
+
+    def has_permission(self, request: request.Request, view) -> bool:
+        if (
+            request.method in CREATE_METHODS
+            and not request.user.is_feature_available("multistructure")
+            and request.user.organization.teams.count() >= 1
+        ):
+            return False
+        return True
+
+
+class OrganizationSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Organization
+        fields = [
+            "id",
+            "name",
+            "created_at",
+            "updated_at",
+        ]
+        read_only_fields = [
+            "id",
+            "created_at",
+            "updated_at",
+        ]
+
+    def create(self, validated_data: Dict, *args: Any, **kwargs: Any) -> Organization:
+        serializers.raise_errors_on_nested_writes("create", self, validated_data)
+        request = self.context["request"]
+        with transaction.atomic():
+            organization = Organization.objects.create(**validated_data)
+            OrganizationMembership.objects.create(
+                organization=organization, user=request.user, level=OrganizationMembership.Level.ADMIN
+            )
+            team = Team.objects.create(organization=organization, name="Default")
+            request.user.current_organization = organization
+            request.user.current_team = team
+            request.user.save()
+        return organization
+
+
+class OrganizationViewSet(viewsets.ModelViewSet):
+    PREMIUM_FEATURES = ["multistructure"]
+
+    serializer_class = OrganizationSerializer
+    pagination_class = None
+    permission_classes = [
+        OrganizationMemberPermissions,
+        OrganizationAdminWritePermissions,
+        PremiumMultiorganizationPermissions,
+    ]
+    queryset = Organization.objects.none()
+    lookup_field = "id"
+    ordering_fields = ["created_by"]
+    ordering = ["-created_by"]
+
+    def get_queryset(self) -> QuerySet:
+        return Organization.objects.filter(
+            id__in=OrganizationMembership.objects.filter(user=self.request.user).values("organization_id")
+        )
+
+    def get_object(self):
+        queryset = self.filter_queryset(self.get_queryset())
+        lookup_value = self.kwargs[self.lookup_field]
+        if lookup_value == "@current":
```

Comment:
Q: Why do we have this edge case?

The frontend should already know the relevant organization id right - this seems to result in added complexity everywhere but tests?

Reply:
It's less work frontend side. Though I can see a way for an inconsistency to creep in if someone changes their org in another tab for instance. 🤔 Would you suggest doing away with this?

Reply:
Good point re switching organizations in tabs. If possible, I think it makes sense to remove it unless we plan to make it a API-wide pattern.

Reply:
Definitely planning to make use of this pattern in the `/user/` endpoint: when we remake it with DRF, let's go with `/users/@me` for accessing the currently logged in user. But in the case of `/organizations` this makes less sense.

Reply:
Decided not to go forward with that approach, as the tradeoff of occasional inconsistency here is constants delays in initial page loads, as it takes a non-negligible amount of waiting until `user` is available (and only then could additional organization- or team-specific data be loaded).